test/membrane_rtc_engine/engine_test.exs

```diff
@@ -0,0 +1,324 @@
+        }
+      ]
+
+      media_event =
+        %{
+          type: "join",
+          data: %{
+            relayAudio: true,
+            relayVideo: true,
+            receiveMedia: true,
+            metadata: metadata,
+            tracksMetadata: tracks_metadata
+          }
+        }
+        |> Jason.encode!()
+
+      send(sfu_engine, {:media_event, peer_id, media_event})
+      assert_receive {_from, {:new_peer, ^peer_id, ^metadata, ^tracks_metadata}}
+      send(sfu_engine, {:accept_new_peer, peer_id})
+      assert_receive {_from, {:sfu_media_event, ^peer_id, media_event}}
+
+      assert %{"type" => "peerAccepted", "data" => %{"id" => peer_id, "peersInRoom" => []}} ==
+               Jason.decode!(media_event)
+    end
+  end
+
+  describe "denying a new peer" do
+    test "triggers peerDenied event", %{sfu_engine: sfu_engine} do
+      peer_id = "sample_id"
+
+      metadata = %{
+        "reason" => "bob smells"
+      }
+
+      tracks_metadata = [
+        %{
+          "type" => "audio",
+          "source" => "microphone"
+        },
+        %{
+          "type" => "video",
+          "source" => "camera"
+        }
+      ]
+
+      media_event =
+        %{
+          type: "join",
+          data: %{
+            relayAudio: true,
+            relayVideo: true,
+            receiveMedia: true,
+            metadata: metadata,
+            tracksMetadata: tracks_metadata
+          }
+        }
+        |> Jason.encode!()
+
+      send(sfu_engine, {:media_event, peer_id, media_event})
+      assert_receive {_from, {:new_peer, ^peer_id, ^metadata, ^tracks_metadata}}
+      send(sfu_engine, {:deny_new_peer, peer_id, data: metadata})
+      assert_receive {_from, {:sfu_media_event, ^peer_id, media_event}}
+
+      assert %{"type" => "peerDenied", "data" => %{"reason" => "bob smells"}} ==
+               Jason.decode!(media_event)
+    end
+  end
+
+  describe "updating peer metadata" do
+    setup do
+      peer_id = "sample_id"
+
+      state = %{
+        id: "room_id",
+        peers: %{
+          peer_id => %{
+            id: peer_id,
+            metadata: %{"displayName" => "Bob"},
+            mid_to_track_metadata: %{
+              "ABC123" => %{"type" => "audio", "source" => "microphone"}
+            }
+          }
+        },
+        incoming_peers: %{},
+        endpoints: %{},
+        options: nil,
+        packet_filters: %{}
+      }
+
+      EngineHelpers.register_test_pid()
+
+      [peer_id: peer_id, state: state]
+    end
+
+    test "triggers peerUpdated event when metadata changes", %{
+      peer_id: peer_id,
+      state: initial_state
+    } do
+      media_event =
+        %{
+          type: "updatePeerMetadata",
+          data: %{
+            metadata: %{"displayName" => "Alice"}
+          }
+        }
+        |> Jason.encode!()
+
+      assert {{:ok, []}, new_state} =
+               Engine.handle_other({:media_event, peer_id, media_event}, %{}, initial_state)
```

Comment:
So good idea to mock engine state in such a way. I will definitely use it in the future

Reply:
Yeah, I started with tests that interact with the running engine, but there were so many messages that would need to be sent, received, then responded to, to put the engine in the state to run.

The added complication was that by the time I set up the engine to unit test, membrane elements deep in the pipeline would start crashing due to the fact that there were no real media tracks streaming.

Reply:
Yeah exactly. This unlocks so many tests scenarios